src/providers/hana/qgshanaconnection.cpp

```diff
@@ -0,0 +1,791 @@
+      values[0].isUnique = values.size() == 1;
+
+    for ( auto lp : values )
+      list << lp;
+  }
+
+  return list;
+}
+
+void QgsHanaConnection::readLayerInfo( QgsHanaLayerProperty &layerProperty )
+{
+  layerProperty.srid = getLayerSRID( layerProperty );
+  layerProperty.type = getLayerGeometryType( layerProperty );
+  layerProperty.pkCols = getLayerPrimaryeKeys( layerProperty );
+}
+
+QVector<QgsHanaSchemaProperty> QgsHanaConnection::getSchemas( const QString &ownerName )
+{
+  QString sql = QStringLiteral( "SELECT SCHEMA_NAME, SCHEMA_OWNER FROM SYS.SCHEMAS WHERE "
+                                "HAS_PRIVILEGES = 'TRUE' AND %1 AND "
+                                "SCHEMA_NAME NOT LIKE_REGEXPR 'SYS|_SYS.*|UIS|SAP_XS|SAP_REST|HANA_XS|XSSQLCC_'" )
+                .arg( !ownerName.isEmpty() ? QStringLiteral( "SCHEMA_OWNER = ?" ) : QStringLiteral( "SCHEMA_OWNER IS NOT NULL" ) );
+
+  QVector<QgsHanaSchemaProperty> list;
+
+  try
+  {
+    PreparedStatementRef stmt = mConnection->prepareStatement( QgsHanaUtils::toUtf16( sql ) );
+    if ( !ownerName.isEmpty() )
+      stmt->setNString( 1, NString( ownerName.toStdU16String() ) );
+    QgsHanaResultSetRef rsSchemas = QgsHanaResultSet::create( stmt );
+    while ( rsSchemas->next() )
+    {
+      QgsHanaSchemaProperty schema;
+      schema.name = rsSchemas->getString( 1 );
+      schema.owner = rsSchemas->getString( 2 );
+      list << schema;
+    }
+    rsSchemas->close();
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+
+  return list;
+}
+
+int QgsHanaConnection::getLayerSRID( const QgsHanaLayerProperty &layerProperty )
+{
+  if ( layerProperty.geometryColName.isEmpty() )
+    return -1;
+
+  int srid = -1;
+
+  try
+  {
+    PreparedStatementRef stmt;
+    if ( !layerProperty.isView )
+    {
+      const char *sql = "SELECT SRS_ID FROM SYS.ST_GEOMETRY_COLUMNS "
+                        "WHERE SCHEMA_NAME = ? AND TABLE_NAME = ? AND COLUMN_NAME = ?";
+      stmt = mConnection->prepareStatement( sql );
+      stmt->setNString( 1, NString( layerProperty.schemaName.toStdU16String() ) );
+      stmt->setNString( 2, NString( layerProperty.tableName.toStdU16String() ) );
+      stmt->setNString( 3, NString( layerProperty.geometryColName.toStdU16String() ) );
+    }
+    else
+    {
+      QString sql = QStringLiteral( "SELECT %1.ST_SRID() FROM %2.%3 WHERE %1 IS NOT NULL LIMIT %4" )
+                    .arg( QgsHanaUtils::quotedIdentifier( layerProperty.geometryColName ),
+                          QgsHanaUtils::quotedIdentifier( layerProperty.schemaName ),
+                          QgsHanaUtils::quotedIdentifier( layerProperty.tableName ),
+                          QString::number( GEOM_TYPE_SELECT_LIMIT ) );
+      stmt = mConnection->prepareStatement( QgsHanaUtils::toUtf16( sql ) );
+    }
+
+    int prevSrid = -1;
+    ResultSetRef rsSrid = stmt->executeQuery( );
+    while ( rsSrid->next() )
+    {
+      srid = *rsSrid->getInt( 1 );
```

Comment:
Once we introduce geometry columns without SRID constraint, the `ST_GEOMETRY_COLUMNS` view could contain `NULL` in the `SRS_ID` field. It probably makes sense to prepare the plugin for that already.

Reply:
Renamed in 2267454.